awx/main/tasks.py

```diff
@@ -2949,18 +2992,112 @@ def __init__(self, task=None, runner_params=None):
             execution_environment_params = self.task.build_execution_environment_params(self.task.instance)
             self.runner_params['settings'].update(execution_environment_params)
 
-    def run(self):
+    def run(self, work_type=None):
         # We establish a connection to the Receptor socket
-        receptor_ctl = ReceptorControl('/var/run/receptor/receptor.sock')
+        receptor_ctl = ReceptorControl(RECEPTOR_SOCK)
 
         try:
-            return self._run_internal(receptor_ctl)
+            return self._run_internal(receptor_ctl, work_type=work_type)
         finally:
             # Make sure to always release the work unit if we established it
             if self.unit_id is not None and settings.RECEPTOR_RELEASE_WORK:
                 receptor_ctl.simple_command(f"work release {self.unit_id}")
 
-    def _run_internal(self, receptor_ctl):
+    @classmethod
+    def check_heartbeat(cls, node):
+        # make a private data dir and env dir
+        private_data_dir = tempfile.mkdtemp(prefix='awx_heartbeat_', dir=settings.AWX_ISOLATION_BASE_PATH)
+        env_path = os.path.join(private_data_dir, 'env')
+        os.makedirs(os.path.join(env_path), mode=0o700)
+        # write a cmdline file for adhoc
+        f = os.fdopen(os.open(os.path.join(env_path, 'cmdline'), os.O_RDWR | os.O_CREAT, stat.S_IREAD | stat.S_IWRITE), 'w')
+        f.write(ansible_runner.utils.args2cmdline('localhost'))
+        f.close()
+        # write a custom facts.d to report the runner version
+        facts_path = os.path.join(private_data_dir, 'facts.d')
+        os.makedirs(facts_path, mode=0o700)
+        with open(os.path.join(facts_path, 'ansible_runner.fact'), 'w') as f:
+            os.chmod(f.name, 0o700)
+            f.write("""#!/usr/bin/env sh\necho "{\\"version\\": \\"`ansible-runner --version`\\"}"\n""")  # noqa
```

Comment:
The combination of this custom executable fact *and* `ansible_processor_vcpus` and `ansible_memtotal_mb` are used to collect heartbeat and capacity metadata, and to replace our previous usage of `psutil`.